ingenialink/ethercat/network.py

```diff
@@ -153,12 +166,60 @@ def disconnect_from_slave(self, servo: EthercatServo) -> None:  # type: ignore [
 
         """
         servo.stop_status_listener()
+        if not self._change_drive_state(servo.slave, pysoem.INIT_STATE):
+            logger.warning("Drive can not reach Init state")
         self.servos.remove(servo)
         if not self.servos:
             self.stop_status_listener()
             self._ecat_master.close()
             self.__is_master_running = False
 
+    def start_pdos(self) -> None:
+        """Configure the PDOs and set slave state to OP for all slaves with mapped PDOs"""
+        self._ecat_master.config_map()
+        self._ecat_master.state = pysoem.SAFEOP_STATE
+        op_servo_list: List[EthercatServo] = []
+        for servo in self.servos:
+            if not (servo._rpdo_maps or servo._tpdo_maps):
+                continue
+            op_servo_list.append(servo)
+            if not self._change_drive_state(servo.slave, pysoem.SAFEOP_STATE):
```

Comment:
This could be highly optimized with a function "_change_driveS_state"
Request the transition to all the applicable slaves, then poll them until all have reached SAFE_OP, instead of doing them sequentially.

Reply:
88889db